mlflow/telemetry/client.py

```diff
@@ -1,33 +1,254 @@
+        self._last_batch_time = time.time()
+
+        try:
+            self._queue.put(batch_records, block=False)
+        except Full:
+            # TODO: record this case
+            _logger.debug("Telemetry queue is full, skipping sending data.")
+
+    def _process_records(self, records: list[TelemetryRecord]):
+        """Process a batch of telemetry records."""
+        try:
+            data = {
+                "records": [
+                    {"data": record.data, "partition-key": record.partition_key}
+                    for record in records
+                ]
+            }
+            response = requests.post(
+                self.telemetry_url,
+                json=data,
+                headers={"Content-Type": "application/json"},
+            )
+            if response.status_code != 200:
+                _logger.debug(
+                    f"Failed to send telemetry records. Status code: {response.status_code}, "
+                    f"Response: {response.text}"
+                )
+        except Exception as e:
+            _logger.debug(
+                f"Failed to process telemetry records. Error: {e}.",
+                exc_info=True,
+            )
+
+    def _consumer_loop(self) -> None:
+        while not self._stop_event.is_set():
+            self._dispatch_task()
+
+        # Drain remaining tasks when stopping
+        while not self._queue.empty():
+            self._dispatch_task()
+
+    def _dispatch_task(self) -> None:
+        """Dispatch a task from the queue to the worker thread pool."""
+        if len(self._active_tasks) >= self._max_workers:
+            _, self._active_tasks = wait(self._active_tasks, return_when=FIRST_COMPLETED)
+
+        try:
+            records = self._queue.get(timeout=1)
+        except Empty:
+            return
 
-    # temp helper function to test
-    # TODO: update this in next PR
-    def add_record(self, record: Record):
+        def _handle(records):
+            self._process_records(records)
+            self._queue.task_done()
+
+        try:
+            future = self._worker_threadpool.submit(_handle, records)
+            self._active_tasks.add(future)
+        except Exception as e:
+            # In case it fails to submit the task to the worker thread pool
+            # such as interpreter shutdown, handle the task in this thread
+            _logger.debug(
+                f"Failed to submit task to worker thread pool. Error: {e}",
+                exc_info=True,
+            )
+            _handle(records)
+
+    def activate(self) -> None:
+        """Activate the async queue to accept and handle incoming tasks."""
+        with self._lock:
+            if self.is_active:
+                return
+
+            self._set_up_threads()
+
+            # Callback to ensure remaining tasks are processed before program exit
+            if not self._atexit_callback_registered:
+                atexit.register(self._at_exit_callback)
```

Comment:
I remember that in databricks notebook REPL the `atexit` hook does not work. (but we can use `sigterm` signal handler instead if needing to make it work in databricks.

Reply:
Does the same thing happen to jupyter notebook? We won't enable it in databricks though

Reply:
need to check. Jupyter might have similar issue.

Reply:
Added a comment to follow up